Net: context cleanup

tbursztyka opened the pull request:
I am not going to manipulate net_context's array into a sys_list. It goes through the array only at one place: to get a free net_context. It's enough as it is, and avoids to add sys_node_t attribute into struct net_context.

Commit:
net/offload: Add dummy functions when offloading is disabled

Thit will help removing usage of #if defined(CONFIG_NET_OFFLOAD) in
relevant places.

Note that static inlines are used instead of #define foo(...), to keep
the parameter check at build time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Reduce the usage of #ifdef/#endif

- Not all #ifdef can be removed: those which have a dedicated attribute
in struct net_context.
- For CONFIG_NET_CONTEXT_CHECK: switching the NET_ASSERT_INFO to
NET_DBG (simpler to read and anyway an error code is returned)

Fixes #8725

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on include/net/net_offload.h:
```diff
@@ -420,6 +422,88 @@ static inline int net_offload_put(struct net_if *iface,
 }
 
+#else
+static inline int net_offload_get(struct net_if *iface,
```
This cannot be done like this. The inline version must be placed together with the documentation block so that documentation is generated properly.

tbursztyka replied:
let's try with @cond INTERNAL_HIDDEN

jukkar replied:
Then we do not get any documentation at all for these functions. We need to follow the documentation pattern like this
```
/**
 * @brief Free/close a network context.
 *
 * @details This releases the context. It is not possible to
 * send or receive data via this context after this call.
 * This is similar as BSD shutdown() function.
 *
 * @param iface Network interface where the offloaded IP stack can be
 * reached.
 * @param context The context to be closed.
 *
 * @return 0 if ok, < 0 if error
 */
static inline int net_offload_put(struct net_if *iface,
				  struct net_context *context)
{
#if defined(CONFIG_NET_OFFLOAD)
	NET_ASSERT(iface);
	NET_ASSERT(net_if_offload(iface));
	NET_ASSERT(net_if_offload(iface)->put);
	return net_if_offload(iface)->put(context);
#else 
	return 0;
#endif 
}
```
This is very ugly looking but needed because of limitations in doxygen. I just do not want to fix this one later when we generate documentation for 2.0.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -119,114 +120,103 @@ int net_context_get(sa_family_t family,
-				return -EINVAL;
+	if (IS_ENABLED(CONFIG_NET_CONTEXT_CHECK)) {
+		if (!IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+			NET_WARN("IPv4 disabled");
```
Why did you remove the asserts and used logging instead? The point in the original code was that the error cases are exceptions and user will not trigger these 99% of the time, so the warning is too much in that case.

tbursztyka replied:
sure it can be NET_DBG

The pb I had with the assert are:

- first the assert itself is a bit heavy as these are non-critical error (unlike a pointer that must be non null for instance. Let's say: the net_pkt pointer all through the recv path).
- second was readability: we do a test, and right below the same test, but reversed, is done for the sake of the assert. In some place it made quite a lot of lines just for the assert, generating a big amount of "useless" lines since we already did a proper test (it could have been a 'false' always put as first parameter to the net assert though)

jukkar replied:
Lets use NET_DBG() then

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -1164,22 +1121,22 @@ static int context_setup_udp_packet(struct net_context *context,
-						  NULL, &addr4->sin_addr);
-	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
-		   net_context_get_family(context) == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) &&
```
This looks like unnecessary change as it just swaps ipv4 and ipv6 blocks.

tbursztyka replied:
read the commit message. 75% of the context.c have this order: af_inet6 then af_inet, and while I made the net_pkt/net_context api changed, I reversed that order in some place. So I decided to move it back.

jukkar replied:
There was just too many changes here that were not related -> more difficult to review. I have no issue if we reverse the if's, just put it to separate commit.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -1216,12 +1173,12 @@ static void context_finalize_packet(struct net_context *context,
-		net_ipv4_finalize(pkt, net_context_get_ip_proto(context));
-	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
-		   net_context_get_family(context) == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) &&
```
ditto

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -1289,26 +1246,26 @@ static int context_sendto(struct net_context *context,
 		return -EDESTADDRREQ;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
```
ditto

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -1493,12 +1450,12 @@ int net_context_send(struct net_context *context,
 		goto unlock;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
```
and here too

Commit:
net/ip: Make struct net_addr holding uninon of ipv6/4 unconditionally

It will help to use IS_ENABLED in place of #ifdef in relevant place.

Only struct net_if uses this structure.
In case only IPv4 is used, it will bloat up this struct by 12 bytes.

There are few reasons why this is "ok" in this case:

- On limited rom/ram system it will be unlikely to find a lot of
network interfaces so it should not harm much to raise the size of
struct net_addr.
- If IPv4 is the only enabled IP version, it gains a good amount of
rom/ram to discard IPv6 support so it is fine to steal a bit of this
gain to bloat up a bit struct net_addr.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on include/net/net_ip.h:
```diff
@@ -278,12 +278,8 @@ struct sockaddr_storage {
 struct net_addr {
```
Typo in commit subject s/uninon/union/

I agree the network code looks a bit ugly because of ifdef's that are needed in the code. But if we try to limit the ram usage, then what you are proposing is no good. Why would we relax our own rules here, I think we should leave the ifdef's here.

tbursztyka replied:
Myeah I am also not completely convinced on that move, though as mentioned in the commit it's a really a little bloat.

Commit:
net/context: Reordering some ipv4/6 logic to follow all other places

During net_pkt/net_context API changes, some ip handling blocks were
ordered ipv4 first, ipv6 second. While it is the contrary everywhere
else. So reordering to get things consistent.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar reviewed (approved):
LGTM
